### PR TITLE
feat: redirect plugin support regex

### LINKED ISSUE #3722: allow using regex in the redirect plugin
### Issue description
I can't find how to config apisix to support regular matching in uri. The nginx configuration is as follows:
```
location ~* /lib/([a-fA-F0-9]+)$ {
    return 301 https://example.com/lib/all/$1;
}
```

### Environment

* apisix version (cmd: `apisix version`):2.2
* OS (cmd: `uname -a`):ubantu
* OpenResty / Nginx version (cmd: `nginx -V` or `openresty -V`):
* etcd version, if have (cmd: run `curl http://127.0.0.1:9090/v1/server_info` to get the info from server-info API):
* apisix-dashboard version, if have:2.3

### PR DESCRIPTION
**Pull request by nanamikon**
### What this PR does / why we need it:

Redirect plugin support regex, similar to the regex_uri in the proxy_rewrite_plugin
Fix #3722 and #4140

### Pre-submission checklist:

* [1] Did you explain what problem does this PR solve? Or what new features have been added?
* [1] Have you added corresponding test cases?
* [1] Have you modified the corresponding document?
* [1] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: apisix/plugins/redirect.lua [hunk 1]
```diff
@@ -35,10 +36,26 @@ local schema = {
...
+                description = "regex uri",
+                type = "string",
+            }
+        },
         http_to_https = {type = "boolean"},
     },
     oneOf = {
-        {required = {"uri"}},
+        {
+            anyOf = {
```
Comment (tokers) [inline apisix/plugins/redirect.lua:54]: What about using `oneOf` so `uri` and `regex_uri` cannot co-exist.

File: apisix/plugins/redirect.lua [hunk 2]
```diff
@@ -35,10 +36,26 @@ local schema = {
     properties = {
         ret_code = {type = "integer", minimum = 200, default = 302},
         uri = {type = "string", minLength = 2, pattern = reg},
+        regex_uri = {
+            description = "new uri that substitute from client uri " ..
```
Comment (tokers) [inline apisix/plugins/redirect.lua:40]: term "new uri" is not accurate to describe the `regex_uri` as it's actually a tuple that contains two elements.

File: apisix/plugins/redirect.lua [hunk 3]
```diff
@@ -79,7 +96,21 @@ end
...
+    if not ok then
+        return false, err
+    end
+
+    if conf.regex_uri and #conf.regex_uri > 0 then
+        local _, _, err = re_sub("/fake_uri", conf.regex_uri[1],
+                conf.regex_uri[2], "jo")
+        if err then
+            return false, "invalid regex_uri(" .. conf.regex_uri[1] ..
+                    ", " .. conf.regex_uri[2] .. "): " .. err
```
Comment (tokers) [inline apisix/plugins/redirect.lua:109]: You can use `string.format` to make the error message composition clearer.

File: apisix/plugins/redirect.lua [hunk 4]
```diff
@@ -129,17 +161,35 @@ function _M.rewrite(conf, ctx)
...
+                local msg = "failed to generate new uri by: " .. uri .. err
+                core.log.error(msg)
+                return 500
+            end
+
+            core.response.set_header("Location", new_uri)
+            return ret_code
+        elseif regex_uri then
+            local new_uri, n, err = re_sub(ctx.var.uri, regex_uri[1],
+                    regex_uri[2], "jo")
```
Comment (tokers) [inline apisix/plugins/redirect.lua:177]:
```suggestion
            local new_uri, n, err = re_sub(ctx.var.uri, regex_uri[1],
                                           regex_uri[2], "jo")
```

File: apisix/plugins/redirect.lua [hunk 5]
```diff
@@ -129,17 +161,35 @@ function _M.rewrite(conf, ctx)
...
+            if not new_uri then
+                local msg = "failed to substitute the uri " .. ctx.var.uri ..
```
Comment (tokers) [inline apisix/plugins/redirect.lua:179]: Use `string.format` to make it clearer.

Comment (nanamikon) [reply apisix/plugins/redirect.lua:179]: ok

File: apisix/plugins/redirect.lua (hunk 3 above)
Comment (nanamikon) [reply apisix/plugins/redirect.lua:109]: ok

File: apisix/plugins/redirect.lua (hunk 2 above)
Comment (nanamikon) [reply apisix/plugins/redirect.lua:40]: get it

File: apisix/plugins/redirect.lua (hunk 1 above)
Comment (nanamikon) [reply apisix/plugins/redirect.lua:54]: I think so

**Comment by tokers** [comment]
@nanamikon CI failed, please check it out.

**Comment by Gary-Airwallex** [comment]
> I have fixed it， but how to do with this problem (Semantic Pull Request — add a semantic PR title)

Could it be a missing space between `feat:` and `redirect`? LOL

File: apisix/plugins/redirect.lua [hunk 6]
```diff
@@ -35,10 +37,22 @@ local schema = {
     properties = {
         ret_code = {type = "integer", minimum = 200, default = 302},
         uri = {type = "string", minLength = 2, pattern = reg},
+        regex_uri = {
+            description = "params array for generating new uri that substitute from client uri " ..
+                    "fist param is regular expression, second one is uri template",
```
Comment (Firstsawyou) [inline apisix/plugins/redirect.lua:42]: Bad style, the description below should remain aligned with the top.

File: apisix/plugins/redirect.lua [hunk 7]
```diff
@@ -79,7 +93,22 @@ end
...
 function _M.check_schema(conf)
-    return core.schema.check(schema, conf)
+    local ok, err = core.schema.check(schema, conf)
+    if not ok then
+        return false, err
+    end
+
+    if conf.regex_uri and #conf.regex_uri > 0 then
+        local _, _, err = re_sub("/fake_uri", conf.regex_uri[1],
+                conf.regex_uri[2], "jo")
```
Comment (Firstsawyou) [inline apisix/plugins/redirect.lua:103]: Ditto, please fix similar problems.

File: apisix/plugins/redirect.lua [hunk 8]
```diff
@@ -129,17 +159,34 @@ function _M.rewrite(conf, ctx)
...
-        if not new_uri then
-            core.log.error("failed to generate new uri by: ", uri, " error: ",
-                           err)
-            return 500
+    if ret_code then
+        if uri then
+            local new_uri, err = concat_new_uri(uri, ctx)
+            if not new_uri then
+                local msg = "failed to generate new uri by: " .. uri .. err
+                core.log.error(msg)
```
Comment (Firstsawyou) [inline apisix/plugins/redirect.lua:167]: I think it's better that way.
```suggestion
                core.log.error("failed to generate new uri by: " .. uri .. err)
```

File: apisix/plugins/redirect.lua [hunk 9]
```diff
@@ -17,7 +17,9 @@
 local core = require("apisix.core")
 local tab_insert = table.insert
 local tab_concat = table.concat
+local string_format = string.format
 local re_gmatch = ngx.re.gmatch
+local re_sub      = ngx.re.sub
```
Comment (spacewander) [inline apisix/plugins/redirect.lua:22]: Please remove extra spaces

File: t/plugin/redirect.t [hunk 10]
```diff
@@ -79,7 +79,68 @@ done
 
 
 
-=== TEST 3: add plugin with new uri: /test/add
+=== TEST 3: add plugin with new regex_uri: /test/1 redirect to http://test.com/1
```
Comment (spacewander) [inline t/plugin/redirect.t:82]: Please add new test at the bottom of the file

**Review by spacewander — CHANGES_REQUESTED** [review]

**Review by spacewander — APPROVED** [review]

**Comment by Firstsawyou** [comment]
> Thanks for review, how can I find this bad style before submit a pr? I have tried luacheck and learned from proxy-rewrite.lua, is there any other tools or docs that I can refer to?

You can refer to this doc: https://github.com/apache/apisix/blob/master/CODE_STYLE.md

**Review by Firstsawyou — APPROVED** [review]

File: docs/zh/latest/plugins/redirect.md [hunk 11]
```diff
@@ -29,6 +29,7 @@ URI 重定向插件。
 | ------------- | ------- | ----------- | ------- | ---------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
 | http_to_https | boolean | 可选        | false   |            | 当设置为 `true` 并且请求是 http 时，会自动 301 重定向为 https，uri 保持不变                                                                                                                                                   |
 | uri           | string  | 可选        |         |            | 可以包含 Nginx 变量的 URI，例如：`/test/index.html`, `$uri/index.html`。你可以通过类似于 `$ {xxx}` 的方式引用变量，以避免产生歧义，例如：`${uri}foo/index.html`。若你需要保留 `$` 字符，那么使用如下格式：`/\$foo/index.html` |
+| regex_uri | array[string] | 可选        |         |                   | 转发到上游的新 `uri` 地址, 使用正则表达式匹配来自客户端的uri，当匹配成功后使用模板替换发送重定向到客户端, 未匹配成功时将客户端请求的uri转发至上游。`uri`和`regex_uri`不可以同时存在。例如：["^/iresty/(.*)/(.*)/(.*)","/$1-$2-$3"] 第一个元素代表匹配来自客户端请求的uri正则表达式，第二个元素代表匹配成功后发送重定向到客户端的uri模板。 |
```
Comment (juzhiyuan) [inline docs/zh/latest/plugins/redirect.md:32]:
```suggestion
| regex_uri | array[string] | 可选        |         |                   | 转发到上游的新 `uri` 地址, 使用正则表达式匹配来自客户端的 `uri`，当匹配成功后使用模板替换发送重定向到客户端, 未匹配成功时将客户端请求的 `uri` 转发至上游。`uri` 和 `regex_uri` 不可以同时存在。例如：["^/iresty/(.*)/(.*)/(.*)","/$1-$2-$3"] 第一个元素代表匹配来自客户端请求的 `uri` 正则表达式，第二个元素代表匹配成功后发送重定向到客户端的 `uri` 模板。 |
```

**Review by juzhiyuan — APPROVED** [review]
LGTM from codes logic

**Review by Yiyiyimu — APPROVED** [review]

**Pull request merged**